plutus-core/plutus-core/src/PlutusCore/Bitwise/Logical.hs

```diff
@@ -0,0 +1,461 @@
+          let (bigIx, littleIx) = i `quotRem` 8
+          let flipIx = len - fromIntegral bigIx - 1
+          w8 :: Word8 <- peekByteOff ptr flipIx
+          let toWrite = if b
+                        then Bits.setBit w8 . fromIntegral $ littleIx
+                        else Bits.clearBit w8 . fromIntegral $ littleIx
+          pokeByteOff ptr flipIx toWrite
+
+-- | Byte replication, as per [CIP-122](https://github.com/mlabs-haskell/CIPs/blob/koz/logic-ops/CIP-0122/CIP-0122.md)
+replicateByteString :: Int -> Word8 -> BuiltinResult ByteString
+replicateByteString len w8
+  | len < 0 = do
+      emit "byteStringReplicate: negative length requested"
+      evaluationFailure
+  | otherwise = pure . BS.replicate len $ w8
+
+-- Helpers
+
+{- Note [writeBits and exceptions]
+
+   As `writeBits` allows us to pass a changelist argument of any length, we
+   potentially could have an out-of-bounds index anywhere in the list. As we
+   have to fail on such cases (and report them appropriately), we end up needing
+   _both_ IO (to do mutable things) as well as a way to signal errors. We can
+   do this in two ways:
+
+   1. Pre-scan the changelist for any out-of-bounds indexes, fail if we see any,
+      then apply the necessary changes if no out-of-bounds indexes are found.
+   2. Speculatively allocate the new `ByteString`, then do the changes in the
+      changelist argument one at a time, failing as soon as we see an out-of-bounds
+      index.
+
+  Option 1 would require traversing the changelist argument twice, which is
+  undesirable, which means that option 2 is the more efficient choice. The
+  natural choice for option 2 would be something similar to `ExceptT Int IO`
+  (with the `Int` being an out-of-bounds index). However, we aren't able to do
+  this, as ultimately, `ByteString`s are implemented as `ForeignPtr`s, forcing
+  us to use the following function to interact with them, directly or not:
+
+  withForeignPtr :: forall (a :: Type) . ForeignPtr a -> (Ptr a -> IO b) -> IO b
+
+  Notably, the function argument produces a result of `IO b`, whereas we would
+  need `MonadIO m => m b` instead. This means that our _only_ choice is to
+  use the exception mechanism, either directly or via some wrappers like
+  `MonadUnliftIO`. While this is unusual, and arguably against the spirit of
+  the use of `IO` relative `ByteString` construction, we don't have any other
+  choice. We decided to use the exception mechanism directly, as while
+  `MonadUnliftIO` is a bit cleaner, it ultimately ends up doing the same thing
+  anyway, and this method at least makes it clear what we're doing.
+
+  This doesn't pose any problems from the point of view of Plutus Core, as this
+  exception cannot 'leak'; we handle it entirely within `writeBits`, and no
+  other Plutus Core code can ever see it.
+-}
+newtype WriteBitsException = WriteBitsException Integer
+  deriving stock (Eq, Show)
+
+instance Exception WriteBitsException
+
+{- Note [Loop sectioning]
```

Comment:
Good notes!  This one's particularly helpful when trying to understand the code.